File: src/drivers/DatabaseDriver.ts
export interface QueryResult<T> {
  rows: T[];
  rowCount: number;
}

export abstract class DatabaseDriver {
  abstract connect(): Promise<void>;
  abstract disconnect(): Promise<void>;
  abstract query<T = Record<string, unknown>>(sql: string, params?: unknown[]): Promise<QueryResult<T>>;

  async queryOne<T = Record<string, unknown>>(sql: string, params?: unknown[]): Promise<T | null> {
    const result = await this.query<T>(sql, params);
    return result.rows[0] ?? null;
  }

  async execute(sql: string, params?: unknown[]): Promise<number> {
    const result = await this.query(sql, params);
    return result.rowCount;
  }
}

File: java/conductor-core/src/main/java/com/nouhouari/conductor/drivers/DatabaseDriver.java
package com.nouhouari.conductor.drivers;

import java.util.List;
import java.util.Optional;

/**
 * Java port of src/drivers/DatabaseDriver.ts. Pure plugin contract — no
 * concrete implementation is shipped, matching the TS side (consumers
 * subclass this and register an instance via {@code ConductorWorld.setDb}).
 */
public abstract class DatabaseDriver {

    public record QueryResult<T>(List<T> rows, int rowCount) {
    }

    public abstract void connect();

    public abstract void disconnect();

    public abstract <T> QueryResult<T> query(String sql, Object... params);

    public <T> Optional<T> queryOne(String sql, Object... params) {
        List<T> rows = this.<T>query(sql, params).rows();
        return rows.isEmpty() ? Optional.empty() : Optional.of(rows.get(0));
    }

    public int execute(String sql, Object... params) {
        return query(sql, params).rowCount();
    }
}
